debian/mongo.1
.\" Man page generated from reStructuredText.
.
.TH "MONGO" "1" "Jun 21, 2018" "4.0" "mongerdb-manual"
.SH NAME
monger \- MongerDB Shell
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SS On this page
.INDENT 0.0
.IP \(bu 2
\fI\%Description\fP
.IP \(bu 2
\fI\%Options\fP
.IP \(bu 2
\fI\%Files\fP
.IP \(bu 2
\fI\%Environment\fP
.IP \(bu 2
\fI\%Keyboard Shortcuts\fP
.IP \(bu 2
\fI\%Use\fP
.UNINDENT
.SH DESCRIPTION
.sp
\fI\%monger\fP is an interactive JavaScript shell interface to
MongerDB, which provides a powerful interface for systems
administrators as well as a way for developers to test queries and
operations directly with the database. \fI\%monger\fP also provides
a fully functional JavaScript environment for use with a MongerDB. This
document addresses the basic invocation of the \fI\%monger\fP shell
and an overview of its usage.
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
Starting in version 4.0, \fI\%monger\fP disables support for TLS 1.0
encryption on systems where TLS 1.1+ is available. For
more details, see 4.0\-disable\-tls\&.
.UNINDENT
.UNINDENT
.SH OPTIONS
.SS Core Options
.INDENT 0.0
.TP
.B \-\-shell
Enables the shell interface. If you invoke the \fBmonger\fP command
and specify a JavaScript file as an argument, or use \fI\%\-\-eval\fP to
specify JavaScript on the command line, the \fI\%\-\-shell\fP option
provides the user with a shell prompt after the file finishes executing.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-nodb
Prevents the shell from connecting to any database instances. Later, to
connect to a database within the shell, see
monger\-shell\-new\-connections\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-norc
Prevents the shell from sourcing and evaluating \fB~/.mongerrc.js\fP on
start up.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-quiet
Silences output from the shell during the connection process.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-port <port>
Specifies the port where the \fBmongerd\fP or \fBmongers\fP
instance is listening. If \fI\%\-\-port\fP is not specified,
\fBmonger\fP attempts to connect to port \fB27017\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-host <hostname>
Specifies the name of the host machine where the
\fBmongerd\fP or \fBmongers\fP is running. If this is not specified,
\fBmonger\fP attempts to connect to a MongerDB process running on
the localhost.
.sp
To connect to a replica set, specify the \fBreplica set name\fP and a seed list of set members. Use the
following form:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
<replSetName>/<hostname1><:port>,<hostname2><:port>,<...>
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For TLS/SSL connections (\fB\-\-ssl\fP), \fBmonger\fP verifies that the
hostname of the \fBmongerd\fP or \fBmongers\fP to which you are connecting matches
the CN or SAN of the \fBmongerd\fP or \fBmongers\fP’s \fB\-\-sslPEMKeyFile\fP certificate.
If the hostname does not match the CN/SAN, \fBmonger\fP will fail to
connect.
.sp
For \fI\%DNS seedlist connections\fP, specify the connection protocol as
\fBmongerdb+srv\fP, followed by the DNS SRV hostname record and any
options. The  \fBauthSource\fP and \fBreplicaSet\fP options, if included in
the connection string, will override any corresponding DNS\-configured options
set in the TXT record. Use of the \fBmongerdb+srv:\fP connection string implicitly
enables TLS/SSL (normally set with \fBssl=true\fP) for the client connection. The
TLS/SSL option can be turned off by setting \fBssl=false\fP in the query string.
.sp
Example:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
mongerdb+srv://server.example.com/?connectionTimeout=3000ms
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
New in version 3.6.

.UNINDENT
.INDENT 0.0
.TP
.B \-\-eval <javascript>
Evaluates a JavaScript expression that is specified as an argument.
\fBmonger\fP does not load its own environment when evaluating code.
As a result many options of the shell environment are not available.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-username <username>, \-u <username>
Specifies a username with which to authenticate to a MongerDB database
that uses authentication. Use in conjunction with the \fB\-\-password\fP and
\fB\-\-authenticationDatabase\fP options.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-password <password>, \-p <password>
Specifies a password with which to authenticate to a MongerDB database
that uses authentication. Use in conjunction with the \fI\%\-\-username\fP
and \fI\%\-\-authenticationDatabase\fP options. To force \fBmonger\fP to
prompt for a password, enter the \fI\%\-\-password\fP option as the
last option and leave out the argument.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help, \-h
Returns information on the options and use of \fBmonger\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBmonger\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose
Increases the verbosity of the output of the shell during the connection
process.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-networkMessageCompressors <string>
New in version 3.4.

.sp
.INDENT 7.0
Changed in version 3.6: .IP \(bu 2
Add support for zlib compressor.
.IP \(bu 2
Enabled by default. To disable, set to \fBdisabled\fP\&.
.UNINDENT

.sp
Enables network compression for communication between this
\fBmonger\fP shell and:
.INDENT 7.0
.IP \(bu 2
a \fBmongerd\fP instance
.IP \(bu 2
a \fBmongers\fP instance.
.UNINDENT
.sp
\fBIMPORTANT:\fP
.INDENT 7.0
.INDENT 3.5
Messages are compressed when both parties enable network
compression. Otherwise, messages between the parties are
uncompressed.
.UNINDENT
.UNINDENT
.sp
You can specify the following compressors:
.INDENT 7.0
.IP \(bu 2
snappy (Default)
.IP \(bu 2
zlib
.UNINDENT
.sp
If you specify multiple compressors, then the order in which you list
the compressors matter as well as the communication initiator. For
example, if a \fI\%monger\fP shell specifies the following network
compressors \fBzlib,snappy\fP and the \fBmongerd\fP specifies
\fBsnappy,zlib\fP, messages between \fI\%monger\fP shell and
\fBmongerd\fP uses \fBzlib\fP\&.
.sp
If the parties do not share at least one common compressor, messages
between the parties are uncompressed. For example, if a
\fI\%monger\fP shell specifies the network compressor
\fBzlib\fP and \fBmongerd\fP specifies \fBsnappy\fP, messages
between \fI\%monger\fP shell and \fBmongerd\fP are not compressed.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ipv6
\fIRemoved in version 3.0.\fP
.sp
Enables IPv6 support and allows \fBmonger\fP to connect to the
MongerDB instance using an IPv6 network. Prior to MongerDB 3.0, you
had to specify \fI\%\-\-ipv6\fP to use IPv6. In MongerDB 3.0 and later, IPv6
is always enabled.
.UNINDENT
.INDENT 0.0
.TP
.B <db name>
Specifies the name of the database to connect to. For
example:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
monger admin
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The above command will connect the \fBmonger\fP shell to the
admin database of the MongerDB deployment running on the local machine. You may specify a remote
database instance, with the resolvable hostname or IP address. Separate
the database name from the hostname using a \fB/\fP character. See the
following examples:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
monger mongerdb1.example.net/test
monger mongerdb1/admin
monger 10.8.8.10/test
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This syntax is the \fIonly\fP way to connect to a specific database.
.sp
To specify alternate hosts and a database, you must use this syntax and cannot
use \fI\%\-\-host\fP or \fI\%\-\-port\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-enableJavaScriptJIT
New in version 4.0.

.sp
Enable the JavaScript engine’s JIT compiler.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-disableJavaScriptJIT

debian/changelog
mongerdb (2.5.4) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/12898

 -- Ernie Hershey <ernie.hershey@mongerdb.com>  Thu, 19 Dec 2013 00:41:46 -0500

mongerdb (2.1.2) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10894

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jun 2012 16:56:28 -0500

mongerdb (1.9.2) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Thu, 11 Aug 2011 16:56:28 -0500

mongerdb (1.9.1) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jul 2011 16:56:28 -0500

mongerdb (1.9.0) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10232

 -- Richard Kreuter <richard@10gen.com>  Sun, 01 May 2011 16:56:28 -0500

mongerdb (1.8.0-rc0) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10245

 -- Richard Kreuter <richard@10gen.com>  Tue, 22 Feb 2011 16:56:28 -0500

mongerdb (1.7.6) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10223

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongerdb (1.7.5) unstable; urgency=low

  * durability journalingin place
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10214

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongerdb (1.7.4) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10210

 -- Richard Kreuter <richard@10gen.com>  Mon, 20 Dec 2010 16:56:28 -0500

mongerdb (1.7.3) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10195

 -- Richard Kreuter <richard@10gen.com>  Tue, 16 Nov 2010 16:56:28 -0500

mongerdb (1.7.2) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10194

 -- Richard Kreuter <richard@10gen.com>  Fri, 29 Oct 2010 16:56:28 -0500

mongerdb (1.7.1) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10193

 -- Richard Kreuter <richard@10gen.com>  Tue, 28 Sep 2010 16:56:28 -0500

mongerdb (1.7.0) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 03 Sep 2010 16:56:28 -0500

mongerdb (1.5.8) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Tue, 03 Aug 2010 16:56:28 -0500

mongerdb (1.5.7) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 30 Jul 2010 16:56:28 -0500


mongerdb (1.5.6) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10179

 -- Richard Kreuter <richard@10gen.com>  Sat, 24 Jul 2010 16:56:28 -0500

mongerdb (1.5.5) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 16 Jul 2010 16:56:28 -0500

mongerdb (1.5.4) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 2 Jul 2010 16:56:28 -0500

mongerdb (1.5.3) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Thu, 17 Jun 2010 16:56:28 -0500

mongerdb (1.5.2) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10143


src/mongo/shell/mongo.js
// monger.js

// Defined in monger.cpp

if (!Monger.prototype) {
    throw Error("Monger.prototype not defined");
}

if (!Monger.prototype.find)
    Monger.prototype.find = function(ns, query, fields, limit, skip, batchSize, options) {
        throw Error("find not implemented");
    };
if (!Monger.prototype.insert)
    Monger.prototype.insert = function(ns, obj) {
        throw Error("insert not implemented");
    };
if (!Monger.prototype.remove)
    Monger.prototype.remove = function(ns, pattern) {
        throw Error("remove not implemented");
    };
if (!Monger.prototype.update)
    Monger.prototype.update = function(ns, query, obj, upsert) {
        throw Error("update not implemented");
    };

if (typeof mongerInject == "function") {
    mongerInject(Monger.prototype);
}

Monger.prototype.setSlaveOk = function(value) {
    if (value == undefined)
        value = true;
    this.slaveOk = value;
};

Monger.prototype.getSlaveOk = function() {
    return this.slaveOk || false;
};

Monger.prototype.getDB = function(name) {
    if ((jsTest.options().keyFile) &&
        ((typeof this.authenticated == 'undefined') || !this.authenticated)) {
        jsTest.authenticate(this);
    }
    // There is a weird issue where typeof(db._name) !== "string" when the db name
    // is created from objects returned from native C++ methods.
    // This hack ensures that the db._name is always a string.
    if (typeof(name) === "object") {
        name = name.toString();
    }
    return new DB(this, name);
};

Monger.prototype._getDatabaseNamesFromPrivileges = function() {
    'use strict';

    const ret = this.adminCommand({connectionStatus: 1, showPrivileges: 1});
    if (!ret.ok) {
        throw _getErrorWithCode(ret, "Failed to acquire database information from privileges");
    }

    const privileges = (ret.authInfo || {}).authenticatedUserPrivileges;
    if (privileges === undefined) {
        return [];
    }

    return privileges
        .filter(function(priv) {
            // Find all named databases in priv list.
            return ((priv.resource || {}).db || '').length > 0;
        })
        .map(function(priv) {
            // Return just the names.
            return priv.resource.db;
        })
        .filter(function(db, idx, arr) {
            // Make sure the list is unique
            return arr.indexOf(db) === idx;
        })
        .sort();
};

Monger.prototype.getDBs = function(driverSession = this._getDefaultSession(),
                                  filter = undefined,
                                  nameOnly = undefined,
                                  authorizedDatabases = undefined) {

    return function(driverSession, filter, nameOnly, authorizedDatabases) {
        'use strict';

        let cmdObj = {listDatabases: 1};
        if (filter !== undefined) {
            cmdObj.filter = filter;
        }
        if (nameOnly !== undefined) {
            cmdObj.nameOnly = nameOnly;
        }
        if (authorizedDatabases !== undefined) {
            cmdObj.authorizedDatabases = authorizedDatabases;
        }

        if (driverSession._isExplicit || !jsTest.options().disableImplicitSessions) {
            cmdObj = driverSession._serverSession.injectSessionId(cmdObj);
        }

        const res = this.adminCommand(cmdObj);
        if (!res.ok) {
            // If "Unauthorized" was returned by the back end and we haven't explicitly
            // asked for anything difficult to provide from userspace, then we can
            // fallback on inspecting the user's permissions.
            // This means that:
            //   * filter must be undefined, as reimplementing that logic is out of scope.
            //   * nameOnly must not be false as we can't infer size information.
            //   * authorizedDatabases must not be false as those are the only DBs we can infer.
            // Note that if the above are valid and we get Unauthorized, that also means
            // that we MUST be talking to a pre-4.0 mongerd.
            //
            // Like the server response mode, this path will return a simple list of
            // names if nameOnly is specified as true.
            // If nameOnly is undefined, we come as close as we can to what the
            // server would return by supplying the databases key of the returned
            // object.  Other information is unavailable.
            if ((res.code === ErrorCodes.Unauthorized) && (filter === undefined) &&
                (nameOnly !== false) && (authorizedDatabases !== false)) {
                const names = this._getDatabaseNamesFromPrivileges();
                if (nameOnly === true) {
                    return names;
                } else {
                    return {
                        databases: names.map(function(x) {
                            return {name: x};
                        }),
                    };
                }
            }
            throw _getErrorWithCode(res, "listDatabases failed:" + tojson(res));
        }

        if (nameOnly) {
            return res.databases.map(function(db) {
                return db.name;
            });
        }

        return res;
    }.call(this, driverSession, filter, nameOnly, authorizedDatabases);
};

Monger.prototype.adminCommand = function(cmd) {
    return this.getDB("admin").runCommand(cmd);
};

/**
 * Returns all log components and current verbosity values
 */
Monger.prototype.getLogComponents = function(driverSession = this._getDefaultSession()) {
    var cmdObj = {getParameter: 1, logComponentVerbosity: 1};
    if (driverSession._isExplicit || !jsTest.options().disableImplicitSessions) {
        cmdObj = driverSession._serverSession.injectSessionId(cmdObj);
    }

    var res = this.adminCommand(cmdObj);
    if (!res.ok)
        throw _getErrorWithCode(res, "getLogComponents failed:" + tojson(res));
    return res.logComponentVerbosity;
};

/**
 * Accepts optional second argument "component",
 * string of form "storage.journaling"
 */
Monger.prototype.setLogLevel = function(
    logLevel, component, driverSession = this._getDefaultSession()) {
    componentNames = [];
    if (typeof component === "string") {
        componentNames = component.split(".");
    } else if (component !== undefined) {
        throw Error("setLogLevel component must be a string:" + tojson(component));
    }
    var vDoc = {verbosity: logLevel};

    // nest vDoc
    for (var key, obj; componentNames.length > 0;) {
        obj = {};
        key = componentNames.pop();
        obj[key] = vDoc;
        vDoc = obj;
    }

    var cmdObj = {setParameter: 1, logComponentVerbosity: vDoc};
    if (driverSession._isExplicit || !jsTest.options().disableImplicitSessions) {
        cmdObj = driverSession._serverSession.injectSessionId(cmdObj);
    }

    var res = this.adminCommand(cmdObj);
    if (!res.ok)
        throw _getErrorWithCode(res, "setLogLevel failed:" + tojson(res));
    return res;
};

Monger.prototype.getDBNames = function() {
    return this.getDBs().databases.map(function(z) {
        return z.name;
    });
};

Monger.prototype.getCollection = function(ns) {
    var idx = ns.indexOf(".");
    if (idx < 0)
        throw Error("need . in ns");
    var db = ns.substring(0, idx);
    var c = ns.substring(idx + 1);
    return this.getDB(db).getCollection(c);
};

Monger.prototype.toString = function() {
    return "connection to " + this.host;
};
Monger.prototype.tojson = Monger.prototype.toString;

/**
 * Sets the read preference.
 *
 * @param mode {string} read preference mode to use. Pass null to disable read
 *     preference.
 * @param tagSet {Array.<Object>} optional. The list of tags to use, order matters.
 *     Note that this object only keeps a shallow copy of this array.
 */
Monger.prototype.setReadPref = function(mode, tagSet) {
    if ((this._readPrefMode === "primary") && (typeof(tagSet) !== "undefined") &&
        (Object.keys(tagSet).length > 0)) {
        // we allow empty arrays/objects or no tagSet for compatibility reasons
        throw Error("Can not supply tagSet with readPref mode primary");
    }
    this._setReadPrefUnsafe(mode, tagSet);
};

// Set readPref without validating. Exposed so we can test the server's readPref validation.
Monger.prototype._setReadPrefUnsafe = function(mode, tagSet) {
    this._readPrefMode = mode;
    this._readPrefTagSet = tagSet;
};

Monger.prototype.getReadPrefMode = function() {
    return this._readPrefMode;
};

Monger.prototype.getReadPrefTagSet = function() {
    return this._readPrefTagSet;
};

// Returns a readPreference object of the type expected by mongers.
Monger.prototype.getReadPref = function() {
    var obj = {}, mode, tagSet;
    if (typeof(mode = this.getReadPrefMode()) === "string") {
        obj.mode = mode;
    } else {
        return null;
    }
    // Server Selection Spec: - if readPref mode is "primary" then the tags field MUST
    // be absent. Ensured by setReadPref.
    if (Array.isArray(tagSet = this.getReadPrefTagSet())) {
        obj.tags = tagSet;
    }

    return obj;
};

/**
 * Sets the read concern.
 *
 * @param level {string} read concern level to use. Pass null to disable read concern.
 */
Monger.prototype.setReadConcern = function(level) {
    if (!level) {
        this._readConcernLevel = undefined;
    } else if (level === "local" || level === "majority") {
        this._readConcernLevel = level;
    } else {
        throw Error("Invalid read concern.");
    }
};

/**
 * Gets the read concern.
 */
Monger.prototype.getReadConcern = function() {
    return this._readConcernLevel;
};

connect = function(url, user, pass) {
    if (url instanceof MongerURI) {
        user = url.user;
        pass = url.password;
        url = url.uri;
    }
    if (user && !pass)
        throw Error("you specified a user and not a password.  " +
                    "either you need a password, or you're using the old connect api");

    // Validate connection string "url" as "hostName:portNumber/databaseName"
    //                                  or "hostName/databaseName"
    //                                  or "databaseName"
    //                                  or full monger uri.
    var urlType = typeof url;
    if (urlType == "undefined") {
        throw Error("Missing connection string");
    }
    if (urlType != "string") {
        throw Error("Incorrect type \"" + urlType + "\" for connection string \"" + tojson(url) +
                    "\"");
    }
    url = url.trim();
    if (0 == url.length) {
        throw Error("Empty connection string");
    }

    if (!url.startsWith("mongerdb://") && !url.startsWith("mongerdb+srv://")) {
        const colon = url.lastIndexOf(":");
        const slash = url.lastIndexOf("/");
        if (url.split("/").length > 1) {
            url = url.substring(0, slash).replace(/\//g, "%2F") + url.substring(slash);
        }
        if (slash == 0) {
            throw Error("Failed to parse mongerdb:// URL: " + url);
        }
        if (slash == -1 && colon == -1) {
            url = "mongerdb://127.0.0.1:27017/" + url;
        } else if (slash != -1) {
            url = "mongerdb://" + url;
        }
    }

    var atPos = url.indexOf("@");
    var protocolPos = url.indexOf("://");
    var safeURL = url;
    if (atPos != -1 && protocolPos != -1) {
        safeURL = url.substring(0, protocolPos + 3) + url.substring(atPos + 1);
    }
    chatty("connecting to: " + safeURL);
    try {
        var m = new Monger(url);
    } catch (e) {
        if (url.indexOf(".mongerdb.net") != -1) {
            print("\n\n*** It looks like this is a MongerDB Atlas cluster. Please ensure that your" +
                  " IP whitelist allows connections from your network.\n\n");
        }

        throw e;
    }
    var db = m.getDB(m.defaultDB);

    if (user && pass) {
        if (!db.auth(user, pass)) {
            throw Error("couldn't login");
        }
    }

    if (_shouldUseImplicitSessions()) {
        chatty("Implicit session: " + db.getSession());
    }

    // Implicit sessions should not be used when opening a connection. In particular, the buildInfo
    // command is erroneously marked as requiring auth in MongerDB 3.6 and therefore fails if a
    // logical session id is included in the request.
    const originalTestData = TestData;
    TestData = Object.merge(originalTestData, {disableImplicitSessions: true});
    try {
        // Check server version
        var serverVersion = db.version();
        chatty("MongerDB server version: " + serverVersion);

        var shellVersion = version();
        if (serverVersion.slice(0, 3) != shellVersion.slice(0, 3)) {
            chatty("WARNING: shell and server versions do not match");
        }
    } finally {
        TestData = originalTestData;
    }

    return db;
};

/** deprecated, use writeMode below
 *
 */
Monger.prototype.useWriteCommands = function() {
    return (this.writeMode() != "legacy");
};

Monger.prototype.forceWriteMode = function(mode) {
    this._writeMode = mode;
};

Monger.prototype.hasWriteCommands = function() {
    var hasWriteCommands = (this.getMinWireVersion() <= 2 && 2 <= this.getMaxWireVersion());
    return hasWriteCommands;
};

Monger.prototype.hasExplainCommand = function() {
    var hasExplain = (this.getMinWireVersion() <= 3 && 3 <= this.getMaxWireVersion());
    return hasExplain;
};

/**
 * {String} Returns the current mode set. Will be commands/legacy/compatibility
 *
 * Sends isMaster to determine if the connection is capable of using bulk write operations, and
 * caches the result.
 */

Monger.prototype.writeMode = function() {

    if ('_writeMode' in this) {
        return this._writeMode;
    }

    // get default from shell params
    if (_writeMode)
        this._writeMode = _writeMode();

    // can't use "commands" mode unless server version is good.
    if (this.hasWriteCommands()) {
        // good with whatever is already set
    } else if (this._writeMode == "commands") {
        this._writeMode = "compatibility";
    }

    return this._writeMode;
};

/**
 * Returns true if the shell is configured to use find/getMore commands rather than the C++ client.
 *
 * Currently, the C++ client will always use OP_QUERY find and OP_GET_MORE.
 */
Monger.prototype.useReadCommands = function() {
    return (this.readMode() === "commands");
};

/**
 * For testing, forces the shell to use the readMode specified in 'mode'. Must be either "commands"
 * (use the find/getMore commands), "legacy" (use legacy OP_QUERY/OP_GET_MORE wire protocol reads),
 * or "compatibility" (auto-detect mode based on wire version).
 */
Monger.prototype.forceReadMode = function(mode) {
    if (mode !== "commands" && mode !== "compatibility" && mode !== "legacy") {
        throw new Error("Mode must be one of {commands, compatibility, legacy}, but got: " + mode);
    }

    this._readMode = mode;
};

/**
 * Get the readMode string (either "commands" for find/getMore commands, "legacy" for OP_QUERY find
 * and OP_GET_MORE, or "compatibility" for detecting based on wire version).
 */
Monger.prototype.readMode = function() {
    // Get the readMode from the shell params if we don't have one yet.
    if (typeof _readMode === "function" && !this.hasOwnProperty("_readMode")) {
        this._readMode = _readMode();
    }

    if (this.hasOwnProperty("_readMode") && this._readMode !== "compatibility") {
        // We already have determined our read mode. Just return it.
        return this._readMode;
    } else {
        // We're in compatibility mode. Determine whether the server supports the find/getMore
        // commands. If it does, use commands mode. If not, degrade to legacy mode.
        try {
            var hasReadCommands = (this.getMinWireVersion() <= 4 && 4 <= this.getMaxWireVersion());
            if (hasReadCommands) {
                this._readMode = "commands";
            } else {
                this._readMode = "legacy";
            }
        } catch (e) {
            // We failed trying to determine whether the remote node supports the find/getMore
            // commands. In this case, we keep _readMode as "compatibility" and the shell should
            // issue legacy reads. Next time around we will issue another isMaster to try to
            // determine the readMode decisively.
        }
    }

    return this._readMode;
};

//
// Write Concern can be set at the connection level, and is used for all write operations unless
// overridden at the collection level.
//

Monger.prototype.setWriteConcern = function(wc) {
    if (wc instanceof WriteConcern) {
        this._writeConcern = wc;
    } else {
        this._writeConcern = new WriteConcern(wc);
    }
};

Monger.prototype.getWriteConcern = function() {
    return this._writeConcern;
};

Monger.prototype.unsetWriteConcern = function() {
    delete this._writeConcern;
};

Monger.prototype.advanceClusterTime = function(newTime) {
    if (!newTime.hasOwnProperty("clusterTime")) {
        throw new Error("missing clusterTime field in setClusterTime argument");
    }

    if (typeof this._clusterTime === "object" && this._clusterTime !== null) {
        this._clusterTime =
            (bsonWoCompare({_: this._clusterTime.clusterTime}, {_: newTime.clusterTime}) >= 0)
            ? this._clusterTime
            : newTime;
    } else {
        this._clusterTime = newTime;
    }
};

Monger.prototype.resetClusterTime_forTesting = function() {
    delete this._clusterTime;
};

Monger.prototype.getClusterTime = function() {
    return this._clusterTime;
};

Monger.prototype.startSession = function startSession(options = {}) {
    // Set retryWrites if not already set on options.
    if (!options.hasOwnProperty("retryWrites") && this.hasOwnProperty("_retryWrites")) {
        options.retryWrites = this._retryWrites;
    }
    const newDriverSession = new DriverSession(this, options);

    // Only log this message if we are running a test
    if (typeof TestData === "object" && TestData.testName) {
        jsTest.log("New session started with sessionID: " +
                   tojsononeline(newDriverSession.getSessionId()) + " and options: " +
                   tojsononeline(options));
    }

    return newDriverSession;
};

Monger.prototype._getDefaultSession = function getDefaultSession() {
    // We implicitly associate a Monger connection object with a real session so all requests include
    // a logical session id. These implicit sessions are intentionally not causally consistent. If
    // implicit sessions have been globally disabled, a dummy session is used instead of a real one.
    if (!this.hasOwnProperty("_defaultSession")) {
        if (_shouldUseImplicitSessions()) {
            try {
                this._defaultSession = this.startSession({causalConsistency: false});
            } catch (e) {
                if (e instanceof DriverSession.UnsupportedError) {
                    chatty("WARNING: No implicit session: " + e.message);
                    this._setDummyDefaultSession();
                } else {
                    print("ERROR: Implicit session failed: " + e.message);
                    throw(e);
                }
            }
        } else {
            this._setDummyDefaultSession();
        }
        this._defaultSession._isExplicit = false;
    }
    return this._defaultSession;
};

Monger.prototype._setDummyDefaultSession = function setDummyDefaultSession() {
    this._defaultSession = new _DummyDriverSession(this);
};

Monger.prototype.isCausalConsistency = function isCausalConsistency() {
    if (!this.hasOwnProperty("_causalConsistency")) {
        this._causalConsistency = false;
    }
    return this._causalConsistency;
};

Monger.prototype.setCausalConsistency = function setCausalConsistency(causalConsistency = true) {
    this._causalConsistency = causalConsistency;
};

Monger.prototype.waitForClusterTime = function waitForClusterTime(maxRetries = 10) {
    let isFirstTime = true;
    let count = 0;
    while (count < maxRetries) {
        if (typeof this._clusterTime === "object" && this._clusterTime !== null) {
            if (this._clusterTime.hasOwnProperty("signature") &&
                this._clusterTime.signature.keyId > 0) {
                return;
            }
        }
        if (isFirstTime) {
            isFirstTime = false;
        } else {
            sleep(500);
        }
        count++;
        this.adminCommand({"ping": 1});
    }
    throw new Error("failed waiting for non default clusterTime");
};

/**
 * Given the options object for a 'watch' helper, determines which options apply to the change
 * stream stage, and which apply to the aggregate overall. Returns two objects: the change
 * stream stage specification and the options for the aggregate command, respectively.
 */
Monger.prototype._extractChangeStreamOptions = function(options) {
    options = options || {};
    assert(options instanceof Object, "'options' argument must be an object");

    let changeStreamOptions = {fullDocument: options.fullDocument || "default"};
    delete options.fullDocument;

    if (options.hasOwnProperty("resumeAfter")) {
        changeStreamOptions.resumeAfter = options.resumeAfter;
        delete options.resumeAfter;
    }

    if (options.hasOwnProperty("startAfter")) {
        changeStreamOptions.startAfter = options.startAfter;
        delete options.startAfter;
    }

    if (options.hasOwnProperty("startAtOperationTime")) {
        changeStreamOptions.startAtOperationTime = options.startAtOperationTime;
        delete options.startAtOperationTime;
    }

    return [{$changeStream: changeStreamOptions}, options];
};

Monger.prototype.watch = function(pipeline, options) {
    pipeline = pipeline || [];
    assert(pipeline instanceof Array, "'pipeline' argument must be an array");

    let changeStreamStage;
    [changeStreamStage, aggOptions] = this._extractChangeStreamOptions(options);
    changeStreamStage.$changeStream.allChangesForCluster = true;
    pipeline.unshift(changeStreamStage);
    return this.getDB("admin")._runAggregate({aggregate: 1, pipeline: pipeline}, aggOptions);
};
